src/connector/src/parser/json_parser.rs

```diff
@@ -84,51 +87,79 @@ impl JsonParser {
-                (msg.record.to_vec(), Op::Insert)
+
+            let mut primary_key = msg.primary_key.to_vec();
+            let mut record = msg.record.to_vec();
+            let key_value = simd_json::to_borrowed_value(&mut primary_key)
+                .map_err(|e| RwError::from(ProtocolError(e.to_string())))?;
+
+            let value_value = if msg.record.is_empty() {
+                None
             } else {
-                (msg.primary_key.to_vec(), Op::Delete)
-            }
-        } else {
-            (payload, Op::Insert)
-        };
-
-        let value: BorrowedValue<'_> = simd_json::to_borrowed_value(&mut payload_mut)
-            .map_err(|e| RwError::from(ProtocolError(e.to_string())))?;
-
-        if let BorrowedValue::Array(ref objects) = value && matches!(op, Op::Insert) {
-            at_least_one_ok(
-                objects
-                    .iter()
-                    .map(|obj| Self::parse_single_value(obj, &mut writer))
-                    .collect_vec(),
-            )
-        } else {
-            let fill_fn = |desc: &SourceColumnDesc| {
-                simd_json_parse_value(
-                    &SourceFormat::Json,
-                    &desc.data_type,
-                    json_object_smart_get_value(&value, desc.name.as_str().into()),
+                Some(
+                    simd_json::to_borrowed_value(&mut record)
+                        .map_err(|e| RwError::from(ProtocolError(e.to_string())))?,
                 )
-                .map_err(|e| {
-                    tracing::error!("failed to process value: {}", e);
-                    e.into()
-                })
             };
-            match op {
-                Op::Insert => writer.insert(fill_fn),
-                Op::Delete => writer.delete(fill_fn),
+            let accessor = UpsertAccess {
+                key_accessor: Some(JsonAccess {
+                    value: Cow::Owned(key_value),
+                    options: Default::default(),
+                }),
+                value_accessor: if msg.record.is_empty() {
+                    None
+                } else {
+                    Some(JsonAccess {
+                        value: Cow::Owned(value_value.unwrap()),
+                        options: Default::default(),
+                    })
+                },
+                primary_key_column_name: String::new(),
+            };
+
+            apply_row_operation_on_stream_chunk_writer(accessor, &mut writer)
+        } else {
+            let value = simd_json::to_borrowed_value(&mut payload)
+                .map_err(|e| RwError::from(ProtocolError(e.to_string())))?;
+            let values = if let simd_json::BorrowedValue::Array(arr) = value {
+                arr
+            } else {
+                vec![value]
+            };
+            let mut errors = Vec::new();
+            let mut guard = None;
+            for value in values {
+                let accessor = UpsertAccess {
```

Comment:
IIUC, this is in the branch when `self.enable_upsert==false`. Then why need a `UpsertAccess` here? Maybe just a `JsonAccess` is enough?

Reply:
The `UpsertAccess` has impl both the ChangeEvent trait and the Access trait. And we need the Access trait to be implemented here. ( And please refresh the page to see the incoming changes.